doc/source/development/rfc/rfc91_dataset_close.rst

```diff
@@ -0,0 +1,184 @@
+
+A new virtual method GDALDataset::Close() is added.
+
+.. code-block:: c++
+
+    /** Do final cleanup before a dataset is destroyed.
+     *
+     * This method is typically called by GDALClose() or the destructor of a
+     * GDALDataset subclass. It might also be called by C++ users before
+     * destroying a dataset. It should not be called on a shared dataset whose
+     * reference count is greater than one.
+     *
+     * It gives a last chance to the closing processus to return an error code if
+     * something goes wrong, in particular in creation / update scenarios where
+     * file write or network communication might occur when finalizing the dataset.
+     *
+     * Implementations should be robust to this method to be called several times
+     * (on subsequent calls, it should do nothing and return CE_None).
+     * Once it has been called, no other method than Close() or the dataset
+     * destructor should be called. RasterBand or OGRLayer owned by the dataset
+     * should be assumed as no longer being valid.
+     *
+     * If a driver implements this method, it should also call it from its
+     * dataset destructor.
+     *
+     * @since GDAL 3.7
+     */
+
+A typical implementation might look as the following:
+
+.. code-block:: c++
+
+    MyDataset::~MyDataset()
+    {
+       MyDataset::Close();
+    }
+
+    CPLErr MyDataset::Close()
+    {
+        CPLErr eErr = CE_None;
+        if( nOpenFlags != OPEN_FLAGS_CLOSED )
+        {
+            if( MyDataset::FlushCache(true) != CE_None )
+                eErr = CE_Failure;
+
+            // Do something driver specific
+            if (m_fpImage != nullptr)
+            {
+                if( VSIFCloseL(m_fpImage) != 0 )
+                {
+                    CPLError(CE_Failure, CPLE_FileIO, "VSIFCloseL() failed");
+                    eErr = CE_Failure;
+                }
+            }
+
+            // Call parent Close() implementation.
+            if( MyParentDatasetClass::Close() != CE_None )
+                eErr = CE_Failure;
+        }
+        return eErr;
+    }
+
+The default GDALDataset::Close() implementation sets nOpenFlags to OPEN_FLAGS_CLOSED
+
+
+C API
+-----
+
+GDALClose() and GDALFlushCache() are modified to return a CPLErr return value.
+
+Backward compatibility
+----------------------
+
+This is an ABI change that should not require more than rebuilding applications
+against GDAL headers.
+
+For out-of-tree drivers that implement GDALDataset::FlushCache(), they need to
+take into account the change in its signature.
+
+Out-of-tree drivers that have write capabilities are also encouraged to
+implement GDALDataset::Close().
+
+Limitations
+-----------
+
+Not all drivers will be modified to implement Close() in the candidate
+implementation, and even those modified might call internal methods of the
+driver that don't do error propagation. Consequently further work might be
```

Comment:
Suggest: drop contraction and add comma

driver that do not do error propagation. Consequently, further work might be

Reply:
done